shell/utility.go

```diff
@@ -128,49 +110,84 @@ func runCommand(command string, state *State, environment []string, workingDirec
 	// ends properly.
 	pwStdout.Close()
 	pwStderr.Close()
-	pwDev3.Close()
+	stdOutput := <-stdoutDoneCh
+	<-stderrDoneCh
+	close(logCh)
 
-	// Cancelling the command may block the pipe reader if the file descriptor
-	// was passed to a child process which hasn't closed it. In this case the
-	// copyOutput goroutine will just hang out until exit.
-	select {
-	case <-copyDoneCh:
-	}
 	log.Printf("-------------------------")
-	log.Printf("[DEBUG] Command execution completed. Reading from output pipe: >&3")
+	log.Printf("[DEBUG] Command execution completed:")
 	log.Printf("-------------------------")
+	o := getOutputMap(stdOutput)
+	newState := NewState(environment, o)
+	log.Printf("[DEBUG] shell script command new state: \"%v\"", newState)
+	return newState, nil
```

Comment:
The revised runCommand() will never return nil (if no state is found in the output, runCommand() will return a State struct where the output element is nil). Whereas the old runCommand() would return nil if there was an error in parseJSON(). The practical effect is that, in the revised version, you no longer get the behaviour where if the create handler doesn't emit state then the read handler is called.